oracle-plugin/src/main/java/io/cdap/plugin/oracle/OracleConnector.java

```diff
@@ -0,0 +1,103 @@
+import io.cdap.cdap.etl.api.batch.BatchSource;
+import io.cdap.cdap.etl.api.connector.Connector;
+import io.cdap.cdap.etl.api.connector.ConnectorSpec;
+import io.cdap.cdap.etl.api.connector.ConnectorSpecRequest;
+import io.cdap.cdap.etl.api.connector.PluginSpec;
+import io.cdap.plugin.common.db.DBConnectorPath;
+import io.cdap.plugin.db.SchemaReader;
+import io.cdap.plugin.db.connector.AbstractDBSpecificConnector;
+import io.cdap.plugin.db.connector.AbstractDBSpecificConnectorConfig;
+import io.cdap.plugin.db.connector.DBSpecificPath;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.mapreduce.lib.db.DBWritable;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A Oracle Database Connector that connects to Oracle database via JDBC.
+ */
+@Plugin(type = Connector.PLUGIN_TYPE)
+@Name(OracleConnector.NAME)
+@Description("This connector creates connections to a Oracle database.")
+@Category("Database")
+public class OracleConnector extends AbstractDBSpecificConnector<OracleSourceDBRecord> {
+  public static final String NAME = "Oracle";
+
+  protected OracleConnector(AbstractDBSpecificConnectorConfig config) {
+    super(config);
+  }
+
+  @Override
+  public boolean supportSchema() {
+    return true;
+  }
+
+  @Override
+  protected Class<? extends DBWritable> getDBRecordType() {
+    return OracleSourceDBRecord.class;
+  }
+
+  protected void setConnectorSpec(ConnectorSpecRequest request, DBConnectorPath path,
+                                  ConnectorSpec.Builder builder) {
+    Map<String, String> properties = new HashMap<>();
+    properties.put(OracleSource.OracleSourceConfig.NAME_USE_CONNECTION, "true");
+    properties.put(OracleSource.OracleSourceConfig.NAME_CONNECTION, request.getConnectionWithMacro());
+    builder.addRelatedPlugin(new PluginSpec(OracleConstants.PLUGIN_NAME, BatchSource.PLUGIN_TYPE, properties));
+
+    String table = path.getTable();
+    if (table == null) {
+      return;
+    }
+
+    properties.put(OracleSource.OracleSourceConfig.IMPORT_QUERY,
+                   String.format("SELECT * FROM %s.%s.%s;", path.getDatabase(), path.getSchema(), table));
+    properties.put(OracleSource.OracleSourceConfig.NUM_SPLITS, "1");
+  }
+
+  @Override
+  protected SchemaReader getSchemaReader() {
+    return new OracleSourceSchemaReader();
+  }
+
+  @Override
+  public StructuredRecord transform(LongWritable longWritable, OracleSourceDBRecord record) {
+    return record.getRecord();
+  }
+
+  @Override
+  protected ResultSet queryDatabases(Connection connection) throws SQLException {
+    return connection.createStatement()
+      .executeQuery(String.format("SELECT NAME AS %s FROM V$DATABASE", RESULTSET_COLUMN_TABLE_CAT));
+  }
+  @Override
+  protected String getTableQuery(DBSpecificPath path) {
+    return String.format("SELECT * from %s.%s", path.getSchema(), path.getTable());
```

Comment:
I see that for oracle you don't pass database here and don't override `getConnection`. How database would be passed to Oracle?
In IMPORT_QUERY you pass database

Reply:
good catch. I should override getConnection for Oracle. also you remind me when generateSpec I should also populate database properties.